### PR TITLE
HSDP Support

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Adds HSDP. Manually tested

### CODE REVIEW

File: composer/trainer/mosaic_fsdp_utils.py [hunk 1]
```diff
@@ -38,7 +38,9 @@
 
 SHARDING_MAP = {
     'NO_SHARD': ShardingStrategy.NO_SHARD,
+    '_HYBRID_SHARD_ZERO2': ShardingStrategy._HYBRID_SHARD_ZERO2,
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp_utils.py:41]: why is this private?

Comment (dakinggg) [reply composer/trainer/mosaic_fsdp_utils.py:41]: fair enough

**Review by dakinggg — DISMISSED** [review]

**Review by dakinggg — COMMENTED** [review]
lgtm, but can this be tested? maybe its not possible with our current CI infra?

File: composer/trainer/mosaic_fsdp_utils.py (hunk 1 above)
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp_utils.py:41]: i assume you need to gate this on some torch version?

**Comment by mvpatel2000** [comment]
> lgtm, but can this be tested? maybe its not possible with our current CI infra?

Manually tested! I don't think with current CI infra

**Review by dakinggg — APPROVED** [review]

File: composer/trainer/mosaic_fsdp_utils.py [hunk 2]
```diff
@@ -38,12 +36,14 @@
 
 SHARDING_MAP = {
     'NO_SHARD': ShardingStrategy.NO_SHARD,
-    '_HYBRID_SHARD_ZERO2': ShardingStrategy._HYBRID_SHARD_ZERO2,
     'SHARD_GRAD_OP': ShardingStrategy.SHARD_GRAD_OP,
-    'HYBRID_SHARD': ShardingStrategy.HYBRID_SHARD,
     'FULL_SHARD': ShardingStrategy.FULL_SHARD,
 }
 
+if using_torch_2():
```
Comment (vchiley) [inline composer/trainer/mosaic_fsdp_utils.py:43]: Does this work for 2.0 or just 2.1?

Comment (mvpatel2000) [reply composer/trainer/mosaic_fsdp_utils.py:43]: It's in 2.0

**Pull request merged**